cli/src/main/java/com/box/l10n/mojito/cli/command/checks/GlossaryTermTrie.java

```diff
@@ -0,0 +1,94 @@
+package com.box.l10n.mojito.cli.command.checks;
+
+import org.apache.commons.lang3.StringUtils;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+public class GlossaryTermTrie {
+
+    private class Node {
+        Map<String, Node> children = new HashMap<>();
+        List<GlossaryTerm> glossaryTerms = new ArrayList<>();
+    }
+
+    private Node rootNode = new Node();
+
+    public GlossaryTermTrie(List<GlossaryTerm> terms) {
+        for(GlossaryTerm term: terms) {
+            this.add(term);
+        }
+    }
+
+    public void add(GlossaryTerm term) {
+        String[] words = term.getTerm().split("\\W+");
+        Node currentNode = rootNode;
+        for (String word : words) {
+            word = word.toLowerCase();
+            if(!currentNode.children.containsKey(word)) {
+                currentNode.children.put(word, new Node());
+            }
+            currentNode = currentNode.children.get(word);
+        }
+        currentNode.glossaryTerms.add(term);
+    }
+
+    public GlossarySearchResult runGlossaryCheck(String source) {
+        String[] words = source.split("\\W+");
+        GlossarySearchResult result = new GlossarySearchResult(source);
+        List<String> failures = new ArrayList<>();
+        for(int i = 0; i < words.length; i++) {
+            Node current = rootNode;
+            for (int j = i; j < words.length; j++) {
+                if (!current.children.containsKey(words[j].toLowerCase())){
+                    break;
+                }
+                current = current.children.get(words[j].toLowerCase());
+                if(!current.glossaryTerms.isEmpty()){
+                    checkGlossaryTerms(source, result, failures, current);
```

Comment:
we found a leaf/match, why do we need to re-process the whole source string in `checkGlossaryTerms`?

could we search in parallel for insensitive and senstive match in the trie, and if one succeed but not the other it is a failure then?

Reply:
The trie uses lower case words as it's key so that we can match words in the source string by setting them to lower case, I don't think we can search in parallel for both cases as we can't predict what case the source string will contain so need to standardise them both to lower case. 

Once we find a match then we need to extract the string as is from the source string and check that it matches the glossary term exactly